doc: revises the namespace specifications and includes some clarifications (#2124)
celestiaorg/celestia-app: specs/src/specs/namespace.md

## Abstract

One of Celestia's core data structures is the namespace.
When a user submits a transaction encapsulating a `MsgPayForBlobs` message to Celestia, they MUST associate each blob with exactly one namespace.
After their transaction has been included in a block, the namespace enables users to take an interest in a subset of the blobs published to Celestia by allowing the user to query for blobs by namespace.

In order to enable efficient retrieval of blobs by namespace, Celestia makes use of a [Namespaced Merkle Tree](https://github.com/celestiaorg/nmt).
See section 5.2 of the [LazyLedger whitepaper](https://arxiv.org/pdf/1905.09274.pdf) for more details.

## Overview

A namespace is composed of two fields: [version](#version) and [id](#id).
A namespace is encoded as a byte slice with the version and id concatenated.

![namespace](./figures/namespace.svg)

### Version

The namespace version is an 8-bit unsigned integer that indicates the version of the namespace.
The version is used to determine the format of the namespace and

> is encoded as a single byte.

Author: The line break is added here because the second part, "is encoded as a single byte," conveys a separate and independent message from the first part. By moving it to the next line, we ensure clarity and emphasize the distinction between the two ideas.

> is encoded as a single byte.

Reviewer: thanks for clarifying! TBH it's not immediately obvious to me when to introduce a semantic line break but I think it's safe to keep one here :)

I guess we could also rewrite this as two sentences:

```
The version is used to determine the format of the namespace.
The version is encoded as a single byte.
```

> is encoded as a single byte.

Author: > TBH it's not immediately obvious to me when to introduce a semantic line break but I think it's safe to keep one here :)

As a general guideline, it is usually recommended to add a line break after each ending period. However, there are other instances where we may want to break a line in the middle. In such cases, a helpful mental model is to view it as commit messages. Similar to keeping each commit short and self-contained, we can break down different semantically independent messages of a line into separate lines, using line breaks. 

> I guess we could also rewrite this as two sentences:

Yes, Your suggested approach also works.

> <!-- It may be useful to indicate the endianness of the encoding) -->

Author: You are right! endianness is for byte order, so would not be applicable on a single byte.

> <!-- It may be useful to indicate the endianness of the encoding) -->

Author: Removed it in https://github.com/celestiaorg/celestia-app/pull/2124/commits/f48db891ca9883db17dfc8285117366377194228.

A new namespace version MUST be introduced if the namespace format changes in a backwards incompatible way.

Note: The `PARITY_SHARE_NAMESPACE` uses the namespace version `255` so that it can be ignored via the `IgnoreMaxNamespace` feature from [nmt](https://github.com/celestiaorg/nmt). The `TAIL_PADDING_NAMESPACE` uses the namespace version `255` so that it remains ordered after all blob namespaces even in the case a new namespace version is introduced.
Below we explain supported user-specifiable namespace versions,
however, we note that Celestia MAY utilize other namespace versions for internal use.
For more details, see the [Reserved Namespaces](#reserved-namespaces) section.

A namespace with version `0` must contain an id with a prefix of 18 leading `0` bytes. The remaining 10 bytes of the id are user-specified.
#### Version 0

The only supported user-specifiable namespace version is `0`.
A namespace with version `0` MUST contain an id with a prefix of 18 leading `0` bytes.
The remaining 10 bytes of the id are user-specified.
Below, we provide examples of valid and invalid encoded user-supplied namespaces with version `0`.

```go
// Valid encoded namespaces

> 0x0000000000000000000000000000000000000000000000000000000001 // transaction namespace

Author: Exactly as Rootul said, all the reserved namespaces are now moved to the `Reserved Namespaces` section.

## Abstract

One of Celestia's core data structures is the namespace. 
When a user submits a transaction encapsulating a `MsgPayForBlobs` message to Celestia, they MUST associate each blob with exactly one namespace. 
After their transaction has been included in a block, the namespace enables users to take an interest in a subset of the blobs published to Celestia by allowing the user to query for blobs by namespace.

In order to enable efficient retrieval of blobs by namespace, Celestia makes use of a [Namespaced Merkle Tree](https://github.com/celestiaorg/nmt). 
See section 5.2 of the [LazyLedger whitepaper](https://arxiv.org/pdf/1905.09274.pdf) for more details.

## Overview

A namespace is composed of two fields: [version](#version) and [id](#id). 
A namespace is encoded as a byte slice with the version and id concatenated. 

![namespace](./figures/namespace.svg)

### Version

The namespace version is an 8-bit unsigned integer that indicates the version of the namespace.
The version is used to determine the format of the namespace and
is encoded as a single byte.

> [//]: # (TODO: It may be useful to indicate the endianness of the encoding)

Author: Relevant issue #2126

0x1000000000000000000000000000000000000000000000000000000000 // invalid because it does not have version 0
0x1111111111111111111111111111111111111111111111111111111111 // invalid because it does not have version 0
```
Any change in the number of leading `0` bytes in the id of a namespace with version `0` is considered a backwards incompatible change and MUST be introduced as a new namespace version.

### ID

The namespace ID is a 28 byte identifier that uniquely identifies a namespace. 
The ID is encoded as a byte slice of length 28. 

> [//]: # (TODO: It may be useful to indicate the endianness of the encoding)

Author: Relevant issue #2126

## Reserved Namespaces
Celestia reserves certain namespaces with specific meanings. 

> Applications SHOULD refrain from using these reserved namespaces for their blob data. [//]: # (TODO: the implication of this is still under investigation, but it is prudent to advise app developers not using this namespace)

Author: Relevant issue #2127

0x1111111111111111111111111111111111111111111111111111111111 // invalid because it does not have version 0
```

Any change in the number of leading `0` bytes in the id of a namespace with version `0` is considered a backwards incompatible change and MUST be introduced as a new namespace version.

### ID

The namespace ID is a 28 byte identifier that uniquely identifies a namespace.
The ID is encoded as a byte slice of length 28.

> <!-- It may be useful to indicate the endianness of the encoding) -->

Author: >  that doesn't inherently represent some value?

Can you provide further clarification on this specific section? 
The interpretation of the namespace impacts in various contexts, particularly in relation to ordering in NMTs. Consequently, the encoding and decoding of the namespace play a crucial role when transmitting data.

> <!-- It may be useful to indicate the endianness of the encoding) -->

Reviewer: AFAIK https://en.wikipedia.org/wiki/Endianness is important to clarify for types like a `int64` but I don't understand how it applies to this `[]byte`

<img width="815" alt="Screenshot 2023-07-19 at 3 05 51 PM" src="https://github.com/celestiaorg/celestia-app/assets/3699047/8a2bf8af-a732-45f8-8074-2096e80ba0d6">

> <!-- It may be useful to indicate the endianness of the encoding) -->

Author: >  You can use the encoding/binary packages in the Go std to read or write multi-byte values from byte slices either in big endian or little endian format. 

Exactly, as soon as we want to read and use that byte slice, then the endianness becomes important. Imagine a namespace query with a namespace encoded in little-endian order (in its Protobuf format), while the receiver expects namespaces in big-endian order, so the query won't be resolved correctly, as there won't be any match in the tree for that namespace due to the mismatch endianness.

## Reserved Namespaces

Celestia reserves certain namespaces with specific meanings.

> Applications SHOULD refrain from using these reserved namespaces for their blob data. <!-- the implication of this is still under investigation, but it is prudent to advise app developers not using this namespace -->

Author: This sentence aims to provide clarity to app developers regarding the consequences of overlooking reserved namespaces and using them for their blobs.

>Isn't it straight up infeasible to use these reserved namespaces

Can you please expand on this part? Why it is infeasible? Is it part of the transaction validity rules? where this check is done? What happens if a blob's namespace has a reserved namespace?

> Applications SHOULD refrain from using these reserved namespaces for their blob data. <!-- the implication of this is still under investigation, but it is prudent to advise app developers not using this namespace -->

Author: Initially, I used `MUST` as well. However, since I was uncertain about whether we strictly prevent such usage and consider blobs with reserved namespaces invalid, I opted to present it as a suggestion using `SHOULD`. I will change it to `MUST` once we get to a final conclusion on this aspect.

> Applications SHOULD refrain from using these reserved namespaces for their blob data. <!-- the implication of this is still under investigation, but it is prudent to advise app developers not using this namespace -->

Author: Ok, found the corresponding part in the code, going to change it to `MUST`.

> Applications SHOULD refrain from using these reserved namespaces for their blob data. <!-- the implication of this is still under investigation, but it is prudent to advise app developers not using this namespace -->

Author: Please see the fix in https://github.com/celestiaorg/celestia-app/pull/2124/commits/f171508ce374deda53303124ca7f6b3235498fed

> Applications SHOULD refrain from using these reserved namespaces for their blob data. <!-- the implication of this is still under investigation, but it is prudent to advise app developers not using this namespace -->

Author: Ok, found the answer in the code, changed `SHOULD` to `MUST` in https://github.com/celestiaorg/celestia-app/pull/2124/commits/f171508ce374deda53303124ca7f6b3235498fed.

> Applications SHOULD refrain from using these reserved namespaces for their blob data. <!-- the implication of this is still under investigation, but it is prudent to advise app developers not using this namespace -->

Author: One question @rootulp: The expected behavior is to reject blob transactions with reserved namespaces? right? asking this for the code audit's sake

> Applications SHOULD refrain from using these reserved namespaces for their blob data. <!-- the implication of this is still under investigation, but it is prudent to advise app developers not using this namespace -->

Reviewer 2: Yes, it would be rejected in ProcessProposal

> Applications SHOULD refrain from using these reserved namespaces for their blob data. <!-- the implication of this is still under investigation, but it is prudent to advise app developers not using this namespace -->

Reviewer: > The expected behavior is to reject blob transactions with reserved namespaces?

Yes.

> Applications SHOULD refrain from using these reserved namespaces for their blob data. <!-- the implication of this is still under investigation, but it is prudent to advise app developers not using this namespace -->

Author: > Yes, it would be rejected in ProcessProposal

If it is checked during `ProcessProposal`, then how would the tx sender be informed that its transaction was invalid? or they are not informed?

> The rationale behind this is that these namespaces dictate the positioning of transactions and blobs data within the Celestia [data square](./data_square_layout.md).

Reviewer: [question] I don't really understand this line. May it be rephrased?

> Therefore, by refraining from using these reserved namespaces, applications can guarantee the desired arrangement of their blob data.

Reviewer: [question] I don't understand this line. May it be rephrased?

Author: I decided to remove this part until the consequences of using reserved namespaces (https://github.com/celestiaorg/celestia-app/issues/2127) become more clear.

Below is a list of reserved namespaces, along with a brief description of each.
In the table, you will notice that the `PARITY_SHARE_NAMESPACE` and `TAIL_PADDING_NAMESPACE` utilize the namespace version `255`, which differs from the supported user-specified versions.

> The reason for employing version `255` for the `PARITY_SHARE_NAMESPACE` is to enable more efficient proof generation within the context of [nmt](https://github.com/celestiaorg/nmt), where it is used in conjunction with the `IgnoreMaxNamespace` feature.

Author: Your point is generally correct. However, I believe you are specifically referring to the aspect of "efficient proof generation." In this case, it differs from simply ordering the shares. When the `IgnoreMaxNamespace` flag is set to true in an NMT, the calculation of parent namespaces from their children excludes the right child if it consists entirely of leaves with the `PARITY_SHARE_NAMESPACE` namespace. In other words, if the minimum and maximum namespaces of the right child are both `PARITY_SHARE_NAMESPACE`, it is ignored. By doing this, the `PARITY_SHARE_NAMESPACE` does not appear in the root of the NMT. As a result, we can immediately respond to queries about parity shares with an empty proof, without the need to search for a match in the tree.

## Assumptions and Considerations

Applications SHOULD refrain from using the [reserved namespaces](#reserved-namespaces) for their blob data.

## Implementation

See [pkg/namespace](../../../pkg/namespace).

## Protobuf Definition

> [//]: # (TODO: Add protobuf definition for namespace if any)

Author: Relevant issue https://github.com/celestiaorg/celestia-app/issues/2128

## Assumptions and Considerations

> Applications SHOULD refrain from using the [reserved namespaces](#reserved-namespaces) for their blob data.

Reviewer: This should be a blob validation rule

```suggestion
Applications MUST NOT use the [reserved namespaces](#reserved-namespaces) for their blob data.
```

## Implementation

See [pkg/namespace](../../../pkg/namespace).

## Protobuf Definition

> <!-- TODO: Add protobuf definition for namespace -->

Author: Not sure either, it is a placeholder just in case we have, or we want to have a protobuf definition. Please see #2128.
